lab/src/Microsoft.ServiceModel.Syndication/src/XmlReaderWrapper.cs

```diff
@@ -0,0 +1,572 @@
+            string result = string.Empty;
+
+            if (await reader.MoveToContentAsync() != XmlNodeType.Element)
+            {
+                //throw new XmlException(Res.Xml_InvalidNodeType, this.NodeType.ToString(), this as IXmlLineInfo);
+                throw new XmlException();
+            }
+
+            if (!reader.IsEmptyElement)
+            {
+                await reader.ReadAsync();
+                result = await ReadStringAsync(reader);
+                if (reader.NodeType != XmlNodeType.EndElement)
+                {
+                    //throw new XmlException(Res.Xml_UnexpectedNodeInSimpleContent, new string[] { this.NodeType.ToString(), "ReadElementString" }, this as IXmlLineInfo);
+                    throw new XmlException();
+                }
+                await reader.ReadAsync();
+            }
+            else
+            {
+                await reader.ReadAsync();
+            }
+
+            return result;
+        }
+
+        private static async Task<string> ReadElementStringAsync(XmlReader reader, string name)
+        {
+            string result = string.Empty;
+
+            if (await reader.MoveToContentAsync() != XmlNodeType.Element)
+            {
+                throw new XmlException("InvalidNodeType");
+            }
+            if (reader.Name != name)
+            {
+                throw new XmlException("ElementNotFound");
+            }
+
+            if (!reader.IsEmptyElement)
+            {
+                result = await ReadStringAsync(reader);
+
+                if (reader.NodeType != XmlNodeType.EndElement)
+                {
+                    throw new XmlException("InvalidNodeType");
+                }
+                await reader.ReadAsync();
+            }
+            else
+            {
+                await reader.ReadAsync();
+            }
+            return result;
+        }
+
+        private static async Task<string> ReadElementStringAsync(XmlReader reader, string localname, string ns)
+        {
+            string result = string.Empty;
+
+            if (await reader.MoveToContentAsync() != XmlNodeType.Element)
+            {
+                throw new XmlException("InvalidNodeType");
+            }
+
+            if (reader.LocalName != localname || reader.NamespaceURI != ns)
+            {
+                throw new XmlException("ElementNotFound");
+            }
+
+            if (!reader.IsEmptyElement)
+            {
+                result = await ReadStringAsync(reader);
+
+                if (reader.NodeType != XmlNodeType.EndElement)
+                {
+                    throw new XmlException("InvalidNodeType");
+                }
+                await reader.ReadAsync();
+            }
+            else
+            {
+                await reader.ReadAsync();
+            }
+            return result;
+        }
+
+        public static async Task WriteNodeAsync(XmlDictionaryWriter writer, XmlReader reader, bool defattr)
+        {
```

Comment:
You've copied the implementation for `WriteNode(**XmlDictionaryReader** reader, bool defattr)` but prototyped it as `WriteNode(**XmlReader** reader, bool defattr)`. XmlReaderWrapper is derived from and wraps an XmlReader (which XmlDictionaryReader derives from) but there's a different implementation that gets used depending on if the reader is the base XmlReader or an XmlDictionaryReader when calling WriteNode.